dvc/dependency/repo.py

```diff
@@ -88,14 +73,13 @@ def dumpd(self):
         return {self.PARAM_PATH: self.def_path, self.PARAM_REPO: self.def_repo}
 
     def download(self, to):
-        with self._make_repo() as repo:
+        cache = self.repo.cache.local
+        with self._make_repo(cache_dir=cache.cache_dir) as repo:
             if self.def_repo.get(self.PARAM_REV_LOCK) is None:
                 self.def_repo[self.PARAM_REV_LOCK] = repo.get_rev()
 
-            cache = self.repo.cache.local
-            with repo.use_cache(cache):
-                _, _, cache_infos = repo.fetch_external([self.def_path])
-            cache.checkout(to.path_info, cache_infos[0])
+            _, _, cache_infos = repo.fetch_external([self.def_path])
+        cache.checkout(to.path_info, cache_infos[0])
```

Comment:
I am not sure if this will cause problems, maybe performance as they use different `State`? cc @efiop

Reply:
Performance - not so much, but `dvc get` on a drive that is on nfs/cifs might hang because of the state, hence why we've used Noop.

Reply:
Plus, seems like this will also not use the cache.type that is configured in the repo for `dvc import`, that's a bigger problem.

Reply:
cache types are with `cache` as it is what is doing checkout. Plus, we can always pass `cache_types` to `external_repo`.

Reply:
Let me explain a bit what I was saying. This change makes the `ExternalRepo/ExternalGitRepo` and all the subrepos it discovers to have separate cache and state instance, but you can set the same cache directory and same cache types beforehand.

For imports, host's cache is doing the checkout, external_repo only downloads it to the cache. So, this works with by just setting `cache_dir` on `external_repo()`.

For get, we use `tmp_dir` and use separate cache types which can be passed to the `external_repo()` and it'll share those cache dir and cache types (but have separate instances of Cache/State).

```python
with external_repo(url, cache_dir="", cache_types=["reflink"]):
    pass
```

^

Reply:
@skshetry Sounds good.